src/node_crypto.cc

```diff
@@ -2579,6 +2580,839 @@ static bool IsSupportedAuthenticatedMode(const EVP_CIPHER_CTX* ctx) {
+      return pkey;
+    }
+  }
+}
+
+static inline Local<Value> BIOToStringOrBuffer(Environment* env,
+                                               BIO* bio,
+                                               PKFormatType format) {
+  BUF_MEM* bptr;
+  BIO_get_mem_ptr(bio, &bptr);
+  if (format == kKeyFormatPEM) {
+    // PEM is an ASCII format, so we will return it as a string.
+    return String::NewFromUtf8(env->isolate(), bptr->data,
+                               NewStringType::kNormal,
+                               bptr->length).ToLocalChecked();
+  } else {
+    CHECK_EQ(format, kKeyFormatDER);
+    // DER is binary, return it as a buffer.
+    return Buffer::Copy(env, bptr->data, bptr->length).ToLocalChecked();
+  }
+}
+
+static bool WritePublicKeyInner(EVP_PKEY* pkey,
+                                const BIOPointer& bio,
+                                const PublicKeyEncodingConfig& config) {
+  if (config.type_.ToChecked() == kKeyEncodingPKCS1) {
+    // PKCS#1 is only valid for RSA keys.
+    CHECK_EQ(EVP_PKEY_id(pkey), EVP_PKEY_RSA);
+    RSAPointer rsa(EVP_PKEY_get1_RSA(pkey));
+    if (config.format_ == kKeyFormatPEM) {
+      // Encode PKCS#1 as PEM.
+      return PEM_write_bio_RSAPublicKey(bio.get(), rsa.get()) == 1;
+    } else {
+      // Encode PKCS#1 as DER.
+      CHECK_EQ(config.format_, kKeyFormatDER);
+      return i2d_RSAPublicKey_bio(bio.get(), rsa.get()) == 1;
+    }
+  } else {
+    CHECK_EQ(config.type_.ToChecked(), kKeyEncodingSPKI);
+    if (config.format_ == kKeyFormatPEM) {
+      // Encode SPKI as PEM.
+      return PEM_write_bio_PUBKEY(bio.get(), pkey) == 1;
+    } else {
+      // Encode SPKI as DER.
+      CHECK_EQ(config.format_, kKeyFormatDER);
+      return i2d_PUBKEY_bio(bio.get(), pkey) == 1;
+    }
+  }
+}
+
+static MaybeLocal<Value> WritePublicKey(Environment* env,
+                                        EVP_PKEY* pkey,
+                                        const PublicKeyEncodingConfig& config) {
+  BIOPointer bio(BIO_new(BIO_s_mem()));
+  CHECK(bio);
+
+  if (!WritePublicKeyInner(pkey, bio, config)) {
+    ThrowCryptoError(env, ERR_get_error(), "Failed to encode public key");
+    return MaybeLocal<Value>();
+  }
+  return BIOToStringOrBuffer(env, bio.get(), config.format_);
+}
+
+static EVPKeyPointer ParsePrivateKey(const PrivateKeyEncodingConfig& config,
+                                     const char* key,
+                                     size_t key_len) {
+  EVPKeyPointer pkey;
+
+  if (config.format_ == kKeyFormatPEM) {
+    BIOPointer bio(BIO_new_mem_buf(key, key_len));
+    CHECK(bio);
+
+    char* pass = const_cast<char*>(config.passphrase_.get());
+    pkey.reset(PEM_read_bio_PrivateKey(bio.get(),
+                                       nullptr,
+                                       PasswordCallback,
+                                       pass));
+  } else {
+    CHECK_EQ(config.format_, kKeyFormatDER);
+
+    if (config.type_.ToChecked() == kKeyEncodingPKCS1) {
```

Comment:
maybe just stylistic, but I'd use a switch() statement for this, and the default would be an assert(), which is basically what you did for the last `else`